server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -1539,31 +1587,88 @@ private record ShardSnapshotMetaDeleteResult(
          *     no further synchronization
          * </p>
          */
-        private final List<ShardSnapshotMetaDeleteResult> shardDeleteResults = new ArrayList<>();
+        private final BytesStreamOutput shardDeleteResults = new ReleasableBytesStreamOutput(bigArrays);
+
+        private int resultCount = 0;
+
+        private final StreamOutput compressed = new OutputStreamStreamOutput(
+            new BufferedOutputStream(
+                new DeflaterOutputStream(Streams.flushOnCloseStream(shardDeleteResults)),
+                DeflateCompressor.BUFFER_SIZE
+            )
+        );
+
+        private final ArrayList<Closeable> resources = new ArrayList<>();
+
+        private final ShardGenerations.Builder shardGenerationsBuilder = ShardGenerations.builder();
+
+        ShardBlobsToDelete() {
+            resources.add(compressed);
+            resources.add(LeakTracker.wrap((Releasable) shardDeleteResults));
+        }
 
         synchronized void addShardDeleteResult(
             IndexId indexId,
             int shardId,
             ShardGeneration newGeneration,
             Collection<String> blobsToDelete
         ) {
-            shardDeleteResults.add(new ShardSnapshotMetaDeleteResult(indexId, shardId, newGeneration, blobsToDelete));
+            try {
+                shardGenerationsBuilder.put(indexId, shardId, newGeneration);
+                new ShardSnapshotMetaDeleteResult(Objects.requireNonNull(indexId.getId()), shardId, blobsToDelete).writeTo(compressed);
```

Comment:
As far as I understand blobsToDelete is already occupy memory we want to optimize. Adding in memory compression will temporary increase heap usage by 20%? How do you know that blobsToDelete will be GC-ed after compression?

Reply:
There's nothing else retaining it apart from the reference in `ShardBlobsToDelete#shardDeleteResults`.